README.rst
=================================
Timefred -- A time tracker for optimizing deep focus.
=================================

``Timefred`` is a CLI app that strives to be the Alfred to your Batman.

Basic usage::

    $ tf on 'unit tests'
    $ tf stop

It also understands "natural language" commands, like::

    $ tf on 'unit tests' 30m ago
Fancy report generation::

    $ tf report

Sync with 3rd-party services::

    $ tf sync
``Timefred`` supports a wide range of functionality, including Jira integration, data aggregation, and daemon mode.

Assumptions
====
- No-one wants to do any time tracking.
- Your boss wants you to do all the time tracking.
- You're Batman.

Like the comic book character, ``Timefred`` works by minimizing the user's interaction with it, while
maximizing distraction-free work, and taking care of all the tedious details.

It does this by {describe 'oh-crap I forgot to toggle my timer'}, zero learning curve, "smart" behavior
(doesn't need to be told to do things) like "on" automatically stops the previous timer etc.

Usage
=====
Here's the minimal usage style::

    $ tf on 'unit tests'
    Start working on 'unit tests'.

    $ tf status
    You have been working on 'unit tests' for less than a minute.

    $ tf stop
    Stopped working on 'unit tests'.

    $ tf report
    Thursday April 05 2021
      ...

``on`` and ``stop`` can take arguments for:

- time (format described further down)
- tag
- note
Example::

    $ tf on research 2 hours ago
    Started working on research at 2:40 PM.

    $ tf s
    You have been working on research for about 2 hours (since 2:40 PM).

    $ tf stop 30 minutes ago
    Stopped working on research at 2:10 PM.

Also illustrating in the previous example is short aliases of all commands,
their first letter. Like, ``s`` for ``status``, ``o`` for ``on``,
``f`` for ``stop``, etc.

Put brief notes on what you've been doing::

    $ tf note waiting for Napoleon to take over the world
    $ tf n another simple note for demo purposes

Tag your activities for fun and profit::

    $ tf tag imp

Get a log of all activities with the ``log`` (or ``l``) command::

    $ tf log

Command reference
=================

Run ``tf -h`` (or ``--help`` or ``help`` or just ``h``)
to get a short command summary of commands.

``on``
------

- Short: ``o``
- Syntax: ``tf (o|on) <name> [<time>...]``

Start tracking time for the project/activity given by `<name>`. For example::

    tf on conquest

tells ``tf`` to start tracking for the activity ``conquest`` *now*.
You can optionally specify a relative time in the past like so::

    tf on conquest 10mins ago

The format of the time is detailed further below.

``stop``
-------

- Short: ``f``
- Syntax: ``tf (f|stop) [<time>...]``

End tracking for the current activity *now*. Just like with ``on`` command
above, you can give an optional time to the past. Example::

    tf stop 10mins ago

tells ``Timefred`` that you finished working on the current activity at, well, 10
minutes ago.

``status``
----------

- Short: ``s``
- Syntax: ``tf (s|status)``

Gives short human-readable message on the current status, i.e., whether anything
is being tracked currently or not. Example::

    $ tf on conqering-the-world
    Start working on conqering-the-world.
    $ tf status
    You have been working on `conqering-the-world` for less than a minute.

``tag``
-------

- Short: ``t``
- Syntax: ``tf (t|tag) <tag>...``

This command adds the given tags to the current activity. Tags are not currently
used within the ``Timefred`` time tracker, but they will be saved in the JSON data
file. You may use them for whatever purposes you like.

For example, if you have a script to generate a HTML report from your ``Timefred``
data, you could tag some activities with a tag like ``red`` or ``important`` so
that activity will appear in red in the final HTML report.

Use it like::

    tf tag red for-joe

adds the tags ``red`` and ``for-joe`` to the current activitiy. You can specify
any number of tags.

Tags are currently for your purpose. Use them as you see fit.

``note``
--------

- Short: ``n``
- Syntax: ``tf (n|note) <note-text>...``

This command adds a note on the current activity. Again, like tags, this has no
significance with the time tracking aspect of ``Timefred``. This is for your own
recording purposes and for the scripts your write to process your ``Timefred`` data.

Use it like::

    tf note Discuss this with the other team.

adds the note ``Discuss this with the other team.`` to the current activity.

``log``
-------

- Short: ``l1``
- Syntax: ``tf (l|log) [today]``

Gives a table like representation of all activities and total time spent on each
of them.

Time format
===========

Currently only the following are recognized. If there is something that is not
handled, but should be, please open an issue about it or a pull request
(function in question is ``parse_time``)

- *n* seconds ago can be written as:
    - *n* seconds ago
    - *n* second ago
    - *n* secs ago
    - *n* sec ago
    - *n* s ago
    - ``a`` in place of *n* in all above cases, to mean 1 second.
    - E.g., ``10s ago``, ``a sec ago`` ``25 seconds ago``, ``25seconds ago``.

- *n* minutes ago can be written as:
    - *n* minutes ago
    - *n* minute ago
    - *n* mins ago
    - *n* min ago
    - ``a`` in place of *n* in all above cases, to mean 1 minute.
    - E.g., ``5mins ago``, ``a minute ago``, ``10 minutes ago``.

- *n* hours ago can be written as:
    - *n* hours ago
    - *n* hour ago
    - *n* hrs ago
    - *n* hr ago
    - ``a`` or ``an`` in place of *n* in all above cases, to mean 1 hour.
    - E.g., ``an hour ago``, ``an hr ago``, ``2hrs ago``.

Where *n* is an arbitrary number and any number of spaces between *n* and the
time unit are allowed (including zero spaces).

Inspiration / Alternatives
=====
While building ``Timefred``, I kept `Watson <https://github.com/TailorDev/Watson>`_ in mind for its impressive functionality, and `ti <https://github.com/richmeta/ti>`_ for its simplicity.

`Watson` was awkward to use, and `ti` lacked some features. I wanted something simpler, and more enjoyable to use, that would just let me do my work and not have to worry about the time tracking.

Status
======

The project is in beta. If you find any bug or have any feedback, please do open
`a GitHub issue <https://github.com/giladbarnea/Timefred/issues>`_.


Installation
======


License
=======

MIT License

Content snippets
===============
# You don't have to do anything
- Why does timeBro save so much time? Because you can completely forget time tracking and do it retrospectively in a relaxed manner. For example just before the end of the working day. Or at the end of the week.
- Instead of constantly having to reflect and interrupt during the day, you make time entries only once using the memory aid - efficiently and in one go.
- There are good reasons why many users prefer to estimate their times at the end of the day rather than using stopwatches. Nobody wants to be constantly interrupted by time tracking, because it costs time and focus.
- Thanks to timeBro, these interruptions of the workflow are eliminated - and even the headache of making estimates.

https://upload.wikimedia.org/wikipedia/en/5/5c/Alfred_Pennyworth_%28Alex_Ross%29.jpg

CHANGES.rst
===============
Release History
===============

.. Changelog entries should follow this format:

   version (release date)
   ======================

   **section**

   - One-line description of change (link to Github issue/PR)

.. Changes should be organized in one of several sections:

   - Added
   - Changed
   - Removed
   - Fixed

0.1.3 (unreleased)
==================



0.1.2 (November 19, 2017)
=========================

This release fixes several Python 3.6 compatibility bugs.

0.1.1 (January 7, 2017)
=======================

This release fixes a few bugs with command line handling.
If you run into any issues with this release, please
`file an issue <https://github.com/giladbarnea/ti/issues>`_!

0.1.0 (June 27, 2016)
=====================

First release of ``ti`` on PyPI!
Thanks to all of the original authors,
particularly Shrikant Sharat,
who has turned over development of ``ti``
to me, Trevor Bekolay.


ti/times.py
import re
from contextlib import suppress
from datetime import datetime, timedelta
from typing import Union

from arrow import Arrow
from arrow.locales import EnglishLocale

from ti.config import config
from ti.util import confirm
from ti.xarrow import XArrow
from ti.xarrow import xarrow_factory

TZINFO = config.time.tz
FORMATS = config.time.formats

ABBREVS = {
    's': 'seconds',
    'm': 'minutes',
    'h': 'hours',
    'd': 'days',
    'w': 'weeks',
    'M': 'months'
    }

DAYS = set(map(str.lower, EnglishLocale.day_abbreviations[1:] + EnglishLocale.day_names[1:]))
"""{'fri', 'friday', ...}"""

NUM2DAY = list(enumerate(map(str.lower, EnglishLocale.day_names[1:]), start=1))
"""[(1, 'monday'), ..., (7, 'sunday')]"""

TIMEUNIT_REMAINDER = "(?:ec(?:ond)?|in(ute)?|(ou)?r|ay|eek|onth)?s?"

HUMAN_RELATIVE = re.compile((rf'(?P<amount1>\d+)\s*(?P<fullunit1>(?P<unit1>([smhdw]))\s*{TIMEUNIT_REMAINDER})\s*'
                             rf'('
                               rf'(?P<amount2>\d+)\s*(?P<fullunit2>(?P<unit2>([smhdw]))\s*{TIMEUNIT_REMAINDER})\s*'
                               rf'('
                                 rf'(?P<amount3>\d+)\s*(?P<fullunit3>(?P<unit3>([smhdw]))\s*{TIMEUNIT_REMAINDER})'
                               rf')?'
                             rf')?'
                             rf'\s*'
                             r'(?:\s+ago\s*)?$'), re.IGNORECASE)
"""3 hours 2m 15 secs ago
Used in `_rel_time2arrow`"""


def now() -> XArrow:
    return xarrow_factory.now(TZINFO)


def isoweekday(day: str) -> int:  # perf: µs
    """
    >>> isoweekday('mon') == 1
    >>> isoweekday('f') == 5
    """
    day = day.lower()
    if len(day) == 1 and day in ('t','s'):
        raise ValueError(f"ambiguous day: {repr(day)} (tuesday/thursday, saturday/sunday)")
    for num, day_name in NUM2DAY:
        if day_name.startswith(day):
            return num
    raise ValueError(f"unknown day: {repr(day)}")


def _day2arrow(day: str) -> XArrow:  # perf: µs
    """
    >>> _day2arrow('thurs')
    <Arrow ...>
    """
    _now = now()
    now_isoweekday = _now.isoweekday()
    day_isoweekday = isoweekday(day)
    diff = abs(now_isoweekday - day_isoweekday)
    if now_isoweekday <= day_isoweekday:
        shift = diff - 7
    else:
        shift = diff * -1
    return _now.shift(days=shift)


def reformat(date: Union[str, Arrow], fmt=FORMATS.date_time) -> str:
    return formatted2arrow(date).format(fmt)


def human2formatted(engtime: str = "now", fmt=FORMATS.date_time) -> str:
    """Called by parse_args(), written as 'start' and 'end' values

    >>> human2formatted("yesterday")
    '04/26/21 17:16:54'
    """
    arrw: Arrow = human2arrow(engtime)
    return arrw.format(fmt)


def _time2replace_dict(time: str) -> dict:
    """
    >>> _time2replace_dict('09:45')
    {'hour': 9, 'minute': 45}
    """
    if '/' in time:
        raise ValueError(f"Don't understand {time = }")
    match = re.match(r'(\d{1,2})(?::(\d{2})(?::(\d{2}))?)?', time)
    if not match:
        raise ValueError(f"Don't understand {time = }")

    hours = int(match.group(1))
    minutes = match.group(2)
    replace = {'hour': hours}
    if minutes is not None:
        replace.update({'minute': int(minutes)})
    else:
        replace.update({'minute': 0})
    seconds = match.group(3)
    if seconds is not None:
        replace.update({'second': int(seconds)})
    else:
        replace.update({'second': 0})
    return replace


def _abs_time2arrow(time: str) -> XArrow:
    """
    >>> _abs_time2arrow('09:45')
    <Arrow ...>
    """
    replace = _time2replace_dict(time)
    return now().replace(**replace)


def _rel_time2arrow(time: str) -> XArrow:
    """
    >>> _rel_time2arrow('3m ago')
    <Arrow ...>
    """
    match = HUMAN_RELATIVE.fullmatch(time)
    if not match:
        raise ValueError(f"Don't understand {time = }")
    grpdict = match.groupdict()
    amount1 = int(grpdict['amount1'])
    unit1 = grpdict['unit1']
    fullunit1 = grpdict['fullunit1']
    if unit1 == 'm' and fullunit1 == 'month':
        unit1 = 'M'
    delta = {ABBREVS[unit1]: amount1}
    if amount2 := grpdict.get('amount2'):
        delta.update({ABBREVS[grpdict['unit2']]: int(amount2)})
        if amount3 := grpdict.get('amount3'):
            delta.update({ABBREVS[grpdict['unit3']]: int(amount3)})
    parsed: XArrow = now() - timedelta(**delta)
    return parsed


def arrows2rel_time(late: Arrow, early: Arrow) -> str:
    """
    >>> arrows2rel_time(now(), now().shift(days=-5, minutes=3))
    '4 days, 23 hours & 57 minutes ago'
    """
    # if (late.year != early.year or
    #         late.month != early.month):
    #     raise NotImplemented(f"Can only handle differences in weeks and below")
    secs = int((late - early).total_seconds())
    if not secs:
        return ''
    return secs2human(secs) + ' ago'


def human2arrow(engtime: Union[str, Arrow] = "now") -> XArrow:
    """
    Format is e.g.::

        <number>[ ?]<unit>[ ago]

    unit::

        s[ec[ond[s]]], m[in[ute[s]]], h[[ou]r[s]]

    For example: "1s ago", "2 minutes ago", "3hr"

    Returns:
        datetime: The difference between now and `engtime`.
    """
    if isinstance(engtime, Arrow):
        if not isinstance(engtime, XArrow) \
                and confirm(f'human2arrow({engtime = }) is regular Arrow, debug?'):
            breakpoint()
        return engtime
    engtime = engtime.lower()

    # 'now', 'today', 'yesterday'
    if engtime in ('now', 'today'):
        return now()
    if engtime == 'yesterday':
        return now().shift(days=-1)

    # 'thurs', ...
    with suppress(ValueError):
        return _day2arrow(engtime)

    # '3m'
    with suppress(ValueError):
        return _rel_time2arrow(engtime)

    # '09:45'
    with suppress(ValueError):
        return _abs_time2arrow(engtime)

    # 05/18/21
    if '/' in engtime:
        return formatted2arrow(engtime)

    # 'Wednesday 09:45'
    day, _, time = engtime.partition(' ')
    # TODO: '05/20/21' doesnt work! and enters here
    day = _day2arrow(day)
    replace = _time2replace_dict(time)
    return day.replace(**replace)

    #
    # parsed = parsedate(engtime)
    #
    #
    # if not parsed:
    #     raise BadTime(f"BadTime: human2dt({engtime = })")
    # return parsed
    # if not engtime or engtime.lower().strip() == 'now':
    #     return now
    #
    # match = re.match(r'(\d+)\s*(m|mins?|minutes?)(\s+ago\s*)?$', engtime, re.X)
    # if match is not None:
    #     minutes = int(match.group(1))
    #     return now - timedelta(minutes=minutes)
    #
    # match = re.match(r'(\d+)\s*(h|hrs?|hours?)(\s+ago\s*)?$', engtime, re.X)
    # if match is not None:
    #     hours = int(match.group(1))
    #     return now - timedelta(hours=hours)
    #
    # raise BadTime(f"Don't understand the time {engtime!r}")


def formatted2arrow(date: Union[str, Arrow]) -> XArrow:
    """Called by log() and status() with 'start' and 'end' values.

    >>> formatted2arrow('19/04/21 10:13:11')
    datetime(2021, 4, 19, 10, 13, 11)

    >>> formatted2arrow('19/04/21')
    datetime.datetime(2021, 4, 19, 0, 0)

    >>> formatted2arrow('10:13:11')
    datetime.datetime(2021, 4, 27, 10, 13, 11)
    """
    if isinstance(date, Arrow):
        if not isinstance(date, XArrow) \
                and confirm(f'formatted2arrow(date) is regular Arrow, debug?'):
            breakpoint()
        return date
    if ' ' in date:
        # "19/04/21 10:13:11" → arrow(...)
        return xarrow_factory.get(date, FORMATS.date_time, tzinfo=TZINFO)
    if '/' in date:
        # "19/04/21" → arrow(...)
        if date.count('/') == 1:
            date += '/' + str(datetime.today().year)[-2:]
        return xarrow_factory.get(date, FORMATS.date, tzinfo=TZINFO)

    # "10:13:11" → datetime(...)
    today = datetime.today()
    return XArrow(today.year, today.month, today.day, *map(int, date.split(':')))
    # return dt


def secs2human(secs: int) -> str:
    """
    >>> secs2human(777600)
    '1 week & 2 days'

    >>> secs2human(7201)
    '2 hours & 1 second'
    """
    strings = []
    if secs >= 604800:
        weeks = int(secs // 604800)
        secs -= weeks * 604800
        strings.append(str(weeks) + ' week' + ('s' if weeks > 1 else ''))

    if secs >= 86400:
        days = int(secs // 86400)
        secs -= days * 86400
        strings.append(str(days) + ' day' + ('s' if days > 1 else ''))

    if secs >= 3600:
        hours = int(secs // 3600)
        secs -= hours * 3600
        strings.append(str(hours) + ' hour' + ('s' if hours > 1 else ''))

    if secs >= 60:
        mins = int(secs // 60)
        secs -= mins * 60
        strings.append(str(mins) + ' minute' + ('s' if mins > 1 else ''))

    if secs:
        strings.append(str(secs) + ' second' + ('s' if secs > 1 else ''))

    human = ', '.join(strings)[::-1].replace(',', '& ', 1)[::-1]
    return human


def timegap(start_time: Arrow, end_time: Arrow) -> str:
    # TODO: this is the same as arrows2rel_time and used only by status()
    diff = end_time - start_time

    mins = int(diff.total_seconds() // 60)

    if mins == 0:
        return 'less than a minute'
    elif mins == 1:
        return 'a minute'
    elif mins < 44:
        return f'{mins} minutes'
    elif mins < 89:
        return 'about an hour'
    elif mins < 1439:
        hours = int(mins // 60)
        mins_remainder = int(mins % 60)
        return f'about {hours} hours and {mins_remainder} minutes'
    elif mins < 2519:
        return 'about a day'
    elif mins < 43199:
        return f'about {mins // 1440} days'
    elif mins < 86399:
        return 'about a month'
    elif mins < 525599:
        return f'about {mins // 43200} months'
    else:
        return 'more than a year'
